superset/db_engine_specs/ydb.py

```diff
@@ -0,0 +1,94 @@
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+from __future__ import annotations
+import json
+import logging
+from typing import (
+    Any,
+    TYPE_CHECKING,
+)
+
+from superset.constants import TimeGrain
+from superset.db_engine_specs.base import BaseEngineSpec
+
+if TYPE_CHECKING:
+    from superset.models.core import Database
+
+
+logger = logging.getLogger(__name__)
+
+
+class YDBEngineSpec(BaseEngineSpec):
+    engine = "yql"
+    engine_aliases = {"ydb", "yql+ydb"}
+    engine_name = "YDB"
+
+    default_driver = "ydb"
+
+    sqlalchemy_uri_placeholder = (
+        "ydb://{host}:{port}/{database_name}"
+    )
+
+    encrypted_extra_sensitive_fields = {"$.connect_args.credentials", "$.credentials"}
+
+    supports_file_upload = False
+
+    _time_grain_expressions = {
+        None: "{col}",
+    }
+
+    @staticmethod
+    def update_params_from_encrypted_extra(
+        database: Database,
+        params: dict[str, Any],
+    ) -> None:
+        if not database.encrypted_extra:
+            return
+
+        try:
+            encrypted_extra = json.loads(database.encrypted_extra)
+            connect_args = params.setdefault("connect_args", {})
+
+            if "protocol" in encrypted_extra:
+                connect_args["protocol"] = encrypted_extra.pop("protocol", "grpc")
+
+            if "credentials" in encrypted_extra:
+                credentials_info = encrypted_extra.pop("credentials", {})
+                credentials = None
+                if "username" in credentials_info:
+                    from ydb import StaticCredentials
+                    credentials = StaticCredentials.from_user_password(
+                        credentials_info.pop("username", None),
+                        credentials_info.pop("password", None),
+                    )
+                elif "token" in credentials_info:
+                    from ydb import AccessTokenCredentials
+                    credentials = AccessTokenCredentials(
+                        credentials_info.pop("token", None)
+                    )
+                elif "service_account_json" in credentials_info:
+                    from ydb.iam import ServiceAccountCredentials
+                    sa_json = credentials_info.pop("service_account_json", {})
```

Comment:
Default args would never be used here or on L81 for same reason as above; you can just index it since you've checked.

Reply:
fixed